### PR TITLE
test(flink): deep dive on the tests marked for Flink in test_client.py

### PR DESCRIPTION
**Pull request by mfatihaktas**
[Previously](https://github.com/ibis-project/ibis/pull/6920), we made a pass over the tests under `ibis/ibis/backends/tests/` and marked them if they fail for the Flink backend. We are now making the second pass over these tests, and looking deeper into those marked previously. This PR contains the changes for the deep dive on `test_client.py`.

**Edit**: The scope of this PR got extended during the review process and after the merging with https://github.com/ibis-project/ibis/pull/7650. Now contains the following major changes:
- Adds support for creating `temporary view` for in-memory `obj` in `create_table()` and `create_view()`.
  - Raises error when the user-set `temp` argument does not match the supported behavior.
- Adds `rename_table()` for Flink backend.
- Adds unit tests that were previously missing for `create_table/view()` etc.
- Corrects some of the tests with `ibis.memtable()` by setting the backend using monkeypatch.

### CODE REVIEW

File: ibis/backends/flink/ddl.py
Comment (deepyaman) [inline ibis/backends/flink/ddl.py:175]:
```suggestion
            return "CREATE TEMPORARY VIEW"
```

File: ibis/backends/flink/ddl.py
Comment (deepyaman) [inline ibis/backends/flink/ddl.py:177]:
```suggestion
            return "CREATE VIEW"
```

**Review by deepyaman — COMMENTED** [review]
Why is one of the tests failing, saying you need DuckDB? Haven't looked into it yet, but curious if you have.

File: ibis/backends/flink/ddl.py
Comment (mfatihaktas) [reply ibis/backends/flink/ddl.py]: Thanks, done.

File: ibis/backends/flink/ddl.py
Comment (mfatihaktas) [reply ibis/backends/flink/ddl.py]: Thanks, done.

**Comment by mfatihaktas** [comment]
> Why is one of the tests failing, saying you need DuckDB? Haven't looked into it yet, but curious if you have.

I looked over [this error](https://github.com/ibis-project/ibis/actions/runs/6737030093/job/18313637452#step:13:190), but I could not reproduce it on my machine. I will look more.

File: ibis/backends/flink/__init__.py
Comment (deepyaman) [inline ibis/backends/flink/__init__.py:344]:
```suggestion
```
I think so, based on the comment above:

```
        Note that: in Flink, tables can be either virtual (VIEWS) or regular
        (TABLES).
        VIEWS can be created from an existing Table object, usually the result
        of a Table API or SQL query. TABLES describe external data, such as a
        file, database table, or message queue. In other words, TABLES refer
        explicitly to tables constructed directly from source/sink connectors.
```

File: ibis/backends/flink/__init__.py
Comment (deepyaman) [inline ibis/backends/flink/__init__.py:361]: We had discussed potentially making `tbl_properties` optional, right (not necessarily in this PR)? If that's the case, would you still have this block? I don't see other backends referencing `file:///tmp` or something in a hardcoded way.

File: ibis/backends/flink/__init__.py
Comment (deepyaman) [inline ibis/backends/flink/__init__.py:440]: Where are all these extra arguments coming from? My understanding is that something like `create_view` would provide more-or-less the same interface as other backends.

File: ibis/backends/tests/test_client.py
Comment (deepyaman) [inline ibis/backends/tests/test_client.py:222]:
```suggestion
```
Unnecessary diff

File: ibis/backends/tests/test_client.py
Comment (deepyaman) [inline ibis/backends/tests/test_client.py:325]: This is problematic, because it won't pass CI. We don't currently have an issue when running something in a single file causes that issue.

File: ibis/backends/flink/__init__.py
Comment (deepyaman) [inline ibis/backends/flink/__init__.py:438]: Is there a reason you do everything with SQL, instead of some of the PyFlink functions? Not sure if something you and @chloeh13q discussed/based on how she's been implementing things, or if it's based on some investigation--just curious.

File: ibis/backends/flink/__init__.py
Comment (deepyaman) [reply ibis/backends/flink/__init__.py:440]: It would be helpful for me if you can point to a backend you may have referenced doing something like this off of; I'm just curious if other backends have different backend-specific `create_view` signatures, or just this one. I did see that the test code is unchanged, so that's good.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:361]: Good question. I have actually spent quite some time looking for a way to create a table without the `connector` option, but could not find any way. In my current understanding, tables have to be based on a "persistent" storage backend. There does not seem to be a "special connector" to data in memory. The closest to this is the [managed table](https://nightlies.apache.org/flink/flink-table-store-docs-release-0.1/docs/development/overview/#managed-table) as far as I could find.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:440]: > Where are all these extra arguments coming from?

I added the extra arguments according to the [doc](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/create/#create-view) for `create view`.

> It would be helpful for me if you can point to a backend you may have referenced doing something like this off of ...

I went over the definition of `create_view` for other backends, but could not spot any other with different signature.
I added the optional parameter `query_expression` per the [doc](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/create/#create-view) to make the `create_view` more flexible, but that parameter is currently not used. I can remove it to make it consistent with other backends. Please let me know.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:438]: I prefer SQL over PyFlink functions because I find the doc for Flink SQL more clear. In my experience, it is much easier to understand what we can do or cannot do with Flink while reading on the SQL doc relative to the PyFlink doc. I have not discussed this with @chloeh13q though.

File: ibis/backends/flink/__init__.py
Comment (chloeh13q) [inline ibis/backends/flink/__init__.py:338]: I think the expected behavior is that "only one of `schema` or `obj` can be provided" (see the docstring)

File: ibis/backends/flink/__init__.py
Comment (chloeh13q) [reply ibis/backends/flink/__init__.py:359]: > I have actually spent quite some time looking for a way to create a table without the connector option, but could not find any way.

Do you mean in Flink? I think it is technically possible, as you mentioned, if the table is a managed one. I'm on the same page as Deepyaman; I'm also not sure that I understand this implementation / agree that this is a clean implementation here.

File: ibis/backends/flink/__init__.py
Comment (chloeh13q) [inline ibis/backends/flink/__init__.py:420]: Missing parameter description for `must_exist`

File: ibis/backends/flink/__init__.py
Comment (chloeh13q) [reply ibis/backends/flink/__init__.py:438]: Yeah I'd vote for removing `query_expression`, at least for the time being anyway. For the other arguments, they sound reasonable to me because I also had to add some arguments that are not in other backends, but I'd just be cautious that the naming of these parameters are consistent across methods (e.g. `if_not_exists`, `must_exist`, etc).

File: ibis/backends/flink/__init__.py
Comment (chloeh13q) [inline ibis/backends/flink/__init__.py:454]: Missing documentation for `temporary`

**Review by chloeh13q — COMMENTED** [review]
Left some comments but the implementation looks mostly reasonable to me. I do see a lot of check warnings that the new lines are not covered by tests; I'd just make sure that everything is properly unit tested, especially where you're introducing different signature for a method in the Flink backend, each parameter should be tested properly to the extent possible (you can write custom backend tests in the `flink` folder, where I added tests for other DDL methods).

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:359]: Thanks. Then we have two options:
1. Revert the change, so we have the following defined in `create_table()`:
```python
        if schema is not None:
            if not tbl_properties:
                raise exc.IbisError(
                    "tbl_properties is required when creating table with schema"
                )
```
which will lead to failures in > 20 tests.

2. Implement the change with [managed table](https://nightlies.apache.org/flink/flink-table-store-docs-release-0.1/docs/development/overview/#managed-table). However, enabling managed table will require additional steps/changes as described [here](https://nightlies.apache.org/flink/flink-table-store-docs-release-0.1/docs/development/overview/#setup-table-store).

@chloeh13q @deepyaman Which option makes more sense to you?

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:438]: Thanks, removed `query_expression`.

> but I'd just be cautious that the naming of these parameters are consistent across methods (e.g. if_not_exists, must_exist, etc).

Good point. I spent some time contemplating on `if_not_exists` vs. `must_exist` vs. `force`. I decided to name these with "most explicit" names possible, but I guess here consistency across functions is more important than being explicit. This is why I now changed them to `force` in the related functions `create_view()`, `rename_table()`, `drop_table()`. Please let me know if you think otherwise.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:338]: Thanks, but there are several places in the `common tests` where both `schema` and `obj` are provided. We need to decide whether to 
(1) check if only one of them is provided and raise an error otherwise, 
(2) support providing both and check if the schemas match. 
Which one makes more sense to you? (The second increases test coverage.)

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:454]: Thanks, added.

**Comment by mfatihaktas** [comment]
> I do see a lot of check warnings that the new lines are not covered by tests; I'd just make sure that everything is properly unit tested, especially where you're introducing different signature for a method in the Flink backend, each parameter should be tested properly to the extent possible (you can write custom backend tests in the `flink` folder, where I added tests for other DDL methods).

@chloeh13q Thanks, great point. I will add these unit tests once I get the approval from you and @deepyaman about the function signatures :)

File: ibis/backends/flink/__init__.py
Comment (chloeh13q) [reply ibis/backends/flink/__init__.py:360]: I agree with @deepyaman .

File: ibis/backends/flink/__init__.py
Comment (chloeh13q) [reply ibis/backends/flink/__init__.py:339]: I checked the implementation in some of the other backends. Did you get this from the BigQuery backend? I guess it's fine if this is how another backend does it

Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:367]: Thanks, implemented `option 1`.

**Comment by mfatihaktas** [comment]
> > Why is one of the tests failing, saying you need DuckDB? Haven't looked into it yet, but curious if you have.
> 
> I looked over [this error](https://github.com/ibis-project/ibis/actions/runs/6737030093/job/18313637452#step:13:190), but I could not reproduce it on my machine.
> 
> The issue is due to `options.default_backend` being `None` in `ibis/config.py: _default_backend()`:
> 
> ```python
> def _default_backend() -> Any:
>     if (backend := options.default_backend) is not None:
>         return backend
> 
>     try:
>         import duckdb as _  # noqa: F401
>     except ImportError:
>         raise com.IbisError(
>             """\
> You have used a function that relies on the default backend, but the default
> backend (DuckDB) is not installed.
> 
> You may specify an alternate backend to use, e.g.
> 
> ibis.set_backend("polars")
> 
> or to install the DuckDB backend, run:
> 
>     pip install 'ibis-framework[duckdb]'
> 
> or
> 
>     conda install -c conda-forge ibis-framework
> 
> For more information on available backends, visit https://ibis-project.org/install
> """
>         )
> 
>     import ibis
> 
>     options.default_backend = con = ibis.duckdb.connect(":memory:")
>     return con
> ```
> 
> At many places in `test_client.py`, `default_backend` is set with monkeypatch, e.g.,
> 
> ```python
> monkeypatch.setattr(ibis.options, "default_backend", ibis.pandas)
> ```
> 
> The issue might be due to test execution order.

@gforsyth Do you have any idea on what might be causing this issue? 
(The [full log](https://github.com/ibis-project/ibis/actions/runs/7066658959/job/19238925729?pr=7479) for reference.)

**Comment by mfatihaktas** [comment]
> > I do see a lot of check warnings that the new lines are not covered by tests; I'd just make sure that everything is properly unit tested, especially where you're introducing different signature for a method in the Flink backend, each parameter should be tested properly to the extent possible (you can write custom backend tests in the `flink` folder, where I added tests for other DDL methods).
> 
> @chloeh13q Thanks, great point. I will add these unit tests once I get the approval from you and @deepyaman about the function signatures :)

@chloeh13q I added unit tests exercising the newly added lines in `test_ddl.py`, but the not-covered-by-tests warnings are still there. I am investigating why this happens, please let me know if you have any idea on why.

**Comment by gforsyth** [comment]
> @gforsyth Do you have any idea on what might be causing this issue? (The [full log](https://github.com/ibis-project/ibis/actions/runs/7066658959/job/19238925729?pr=7479) for reference.)

If you create a `memtable`, by default, Ibis will use `DuckDB` to materialize it, so the first line of the test in question is 

```
t0 = ibis.memtable({"a": [1, 2, 3]})
```

And even though you then pass it to `ddl_con`, it is going to try to first register the `memtable` with the default backend.

You need to set `ibis.set_backend(ddl_con)` before you create the `memtable` (or you can do it with the `monekypatch` line as above).

It's very possible some of these tests are passing "incorrectly" because it's actually running on `DuckDB` (also possible that we wrote some of those incorrect tests)

**Comment by gforsyth** [comment]
> You need to set `ibis.set_backend(ddl_con)` before you create the `memtable` (or you can do it with the `monekypatch` line as above).

sorry, definitely use the `monkeypatch` method so that this change doesn't move outside the scope of this test

File: ibis/backends/flink/__init__.py
Comment (gforsyth) [inline ibis/backends/flink/__init__.py:176]: This is a deviation from the existing API in Ibis.  

Our current convention is to include views in the output of `list_tables`.

File: ibis/backends/flink/__init__.py
Comment (gforsyth) [inline ibis/backends/flink/__init__.py:170]: As my other comment alludes to, we probably want to include the views in the output of `list_tables`, but if this stays in, the docstring should be fixed to refer to `views`

File: ibis/backends/flink/__init__.py
Comment (gforsyth) [inline ibis/backends/flink/__init__.py:376]: There isn't support for creating a `memtable` out of an existing ibis expression, so i don't think you want this here.

File: ibis/backends/flink/__init__.py
Comment (gforsyth) [inline ibis/backends/flink/__init__.py:385]: This is confusing to me.  If a user defines an `UnboundTable` that they would like to create in the Flink backend, they're going to end up with a `view` instead -- I think a call to `create_table` should create a table.

Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:385]: Many thanks @gforsyth for all your comments. They urged me to go back and discuss the behavior of `create_table()` for Flink with @deepyaman (CC: @chloeh13q, @zhenzhongxu). Here is our understanding / conclusion:
* Flink needs a `source connector` to create a (regular) table.
    *  Ref: [Create Table](https://nightlies.apache.org/flink/flink-table-store-docs-master/docs/development/create-table/)
* In-memory data does not have a connector information. This is why, we can only create a `temporary view` in Flink with in-memory data. (`Views` are referred to as `virtual tables` in Flink.)
    * Ref: [Create Tables in the Catalog](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/common/#create-tables-in-the-catalog)
    > Tables can be either virtual (VIEWS) or regular (TABLES). VIEWS can be created from an existing Table object, usually the result of a Table API or SQL query. TABLES describe external data, such as a file, database table, or message queue.

Given these, we have two options in `create_table()` for Flink backend:
1. Require the connector information to be provided. This would imply that we do not support creating table from `in-memory` data (e.g., `pd.Dataframe`, `pa.Table`).
2. Create a view for in-memory data, as currently done.
3. As far as we understand it, `ir.Table` does not have connector information. If we go with option-1, then we won't support creating table from `ir.Table`. If we go with option-2 and `obj` points to an in-memory data (e.g., `obj = ibis.memtable([("a", 1.0)], columns=["a", "b"])`), we create a view. If `obj` is an `UnboundTable`, we do not support it.

Can you please share your opinion and feedback on these conclusions?

Comment (gforsyth) [reply ibis/backends/flink/__init__.py:385]: We'll probably have to revisit this after this PR and the work in #7640 are consolidated.

I think option 2 is fine for now.  I think there should be support for unbound tables, but that may require some notion of a "default" connector, which may not make sense -- that's still unclear to me.

Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: Thanks @gforsyth. I implemented option 2 and consolidated my changes with those in https://github.com/ibis-project/ibis/pull/7640.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: Thanks, included `views` in `list_tables()`.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: Thanks, included views in `list_tables()`.

**Review by zhenzhongxu — COMMENTED** [review]
Thanks for putting this together. I've reviewed mostly the part that consolidates with my other PR. Looks mostly good, had a few questions and nit picks.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:390]: Thanks, removed.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:122]: Thanks, removed the argument `temporary`. Now `list_tables()` returns all tables and view, including the temporary ones.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:159]: Sorry, could not understand what you meant.
- `_list_tables()` returns only tables.
- `_list_views()` returns only views.
- `list_tables()` returns both tables and views.

So, `_list_tables()` does not return views, hence no duplicate would be returned by `list_tables()`. Please let me know if I am missing something.

Comment (zhenzhongxu) [reply ibis/backends/flink/__init__.py:159]: i was referring according to the documentation, that `_list_tables()` currently returns both tables and views.
specifically, this call `self._table_env._j_tenv.listTables()` is the culprit - see the docstring attached above.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:433]: The reason to have different implementations for creating `view` in `create_table()` and `create_view()` is two-fold:
1. In `create_table()`, we create a `temporary view` for in-memory `obj` using PyFlink as
```py
            table = self._table_env.from_pandas(
                dataframe, FlinkRowSchema.from_ibis(schema)
            )
            self._table_env.create_temporary_view(qualified_name, table)
```
where `from_pandas()` implicitly creates an "intermediate" table source in Arrow and bases the PyFlink table on it. As PyFlink takes care of setting up the intermediate arrow source for us, it makes sense to use PyFlink API here.

2. In `create_view()`, I added `CreateView` to construct the SQL for creating view. Alternative would be using PyFlink API (as we did above for 1.), but I could not find a way to `CREATE VIEW AS SELECT` using PyFlink. I also think that the "convention" in `create_table()` with the SQL builder class `CreateTableFromConnector` is more flexible in general as we directly use Flink SQL without PyFlink as an intermediary. Also in my experience, Flink SQL is much better documented than PyFlink API.

Does this reasoning make sense to you?

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: Grepped `create_view`, but could not find any other backend with `temporary`. What is your opinion? Should we remove it from `create_view()` as well?

File: ibis/backends/flink/ddl.py
Comment (mfatihaktas) [reply ibis/backends/flink/ddl.py:43]: Thanks, done.

File: ibis/backends/tests/test_client.py
Comment (mfatihaktas) [reply ibis/backends/tests/test_client.py]: Thanks for catching this, fixed it.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: To create a memory table view with SQL, we would need to first create an arrow source from `obj` as done by PyFlink. We can surely do that by imitating what is being done by PyFlink, if we decide to do and prioritize it.

**Review by gforsyth — CHANGES_REQUESTED** [review]
Hey @mfatihaktas -- sorry for the delay!  I had a few comments pending and I forgot to submit.

I've flagged a few things that I think need fixing.  Also, because this PR has changed so much (which is fine!), I think updating the PR description with what it is actually doing will be very helpful to our future selves.

File: ibis/backends/flink/__init__.py
Comment (gforsyth) [reply ibis/backends/flink/__init__.py:515]: If there are two to make sure the view gets dropped whether it is `temporary` or not, I would suggest updating `_drop_table` so you can pass in an argument to that effect, instead of duplicating the call with a single argument changed.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:128]: Thanks, modified the docstring as you suggested.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: The suggestion to remove `temp` argument was brought up by @zhenzhongxu, which I agreed at the time.
Our rationale was that we can keep the interface simple initially by managing `temp` for the user, and expose it to user later if/when necessary. This train of thought came up after we decided to support creating only `temporary views` for in-memory `obj`. As we already "decide for the user" when `obj` is in-memory, we thought it is perhaps a good idea to eliminate `temp` from the interface.

However, given that `temp` is exposed to user in other backends, it makes sense to keep it for Flink as well. This however requires an additional step in `create_table()` to handle the case when `obj` is `Dataframe` but `temp` is `False`, which is not supported. Currently, I added a "disclaimer" in the docstring of `create_table()` (as well as `create_view()`) to inform the user that we override `temp` as `True` when `obj` is `Dataframe`. Also added a `TODO` there to check with the reviewers if this makes sense or not. Please take a look and let me know what you think.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:403]: Thanks, please take a look at the updated error message.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: Addressed as described [above](https://github.com/ibis-project/ibis/pull/7479#discussion_r1424508857).

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: Removed while bringing `temp` back.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: Removed while bringing `temp` back.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: Addressed while bringing `temp` back.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:371]: Thanks, done.

File: ibis/backends/flink/__init__.py
Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py:372]: I agree that erroring is cleaner and more explicit than overriding, which is actually what I did at first :)
However with erroring, I encountered the following complication: `temp` is `False` by default for all backends and this makes sense to me.  In many places though, `create_table()` is called with in-memory `obj` and without `temp`, which triggers the error for `isinstance(obj, pd.DataFrame) and temp is False`. I could not find a clean way to modify the tests to make them fit this unusual behavior by Flink backend.

I switched to overriding with the following in mind: As discussed at some point, we might add support for creating tables (instead of creating views as we do now) which would enable supporting `temp=False` for in-memory `obj` as well. This would eliminate the need to override `temp`.

Given this background, do you have any other suggestion? Or should I just switch back to erroring (from overriding) and deal with the test failures by marking them for Flink?

Comment (gforsyth) [reply ibis/backends/flink/__init__.py:372]: I would error and mark the tests as `notimpl`, since those tests are for creating non-temp tables.  Then if we figure out a good pattern for persistent table/view creation from in-memory tables, we'll get a nice signal that it's working.

Comment (mfatihaktas) [reply ibis/backends/flink/__init__.py]: Thanks, switched to erroring.

**Review by gforsyth — APPROVED** [review]
Really nice work here, @mfatihaktas ! Thanks for bearing with me through the various rounds of changes and rebasing.

**Pull request merged**
